[WIP] alignment attribute

o-petrov opened the pull request:
Discuss __aligned__ C attribute support.

o-petrov committed:
Add class and stub for alignment attributes and specifier in CType.

If CType represents a variable with specified __attribute__((__aligned__(x))) or _Alignas(x), it has to store it.
Additionally the type itself can be aligned with same attribute.

o-petrov committed:
Replace Integer with int, null with special constant in Alignment

o-petrov committed:
Add alignment to CSimpleType

o-petrov committed:
Get alignment for simple declaration. Use CSimpleType alignment when getting alignof(type).

o-petrov committed:
Add max fundamental alignment to MachineModel

o-petrov committed:
Test toASTString on aligned numeric types

o-petrov committed:
fix Alignment.string...() names

o-petrov committed:
remove spaces in Alignment.string...()

o-petrov committed:
add alignment in CPointerType

o-petrov committed:
add tests for aligned pointers

o-petrov committed:
parse alignment in CPointerType

o-petrov committed:
use alignment of CPointerType in MachineModel

o-petrov committed:
change CSimpleType.toASTString(): print aligned attribute after declarator instead of between type and declarator

o-petrov committed:
add alignment in CTypedefType

o-petrov committed:
CTypedefType.toASTString: remove space before empty declarator

o-petrov committed:
Replace CTypes.withAlignment with overrideAlignment. Add updateAlignment, as in most cases we need to add some kind of alignment without clearing other kinds.

o-petrov committed:
Add alignment to CArrayType

o-petrov committed:
add arrays to CType toString alignment test

o-petrov committed:
Save alignment for address-of, pointer dereference, array subscript, and pointer arithmetic operators

o-petrov committed:
Remove nodes from CFA in cycle, not recursively.

For every node that has to be removed with its successors, place successors in queue instead of calling remove function recursively on them.

o-petrov committed:
Refactor CType alignment overriding. Make a function for removing variable alignment attribute and Alignas specifier from Ctype.

o-petrov committed:
Add aligned attribute testing script

Generate programs with expressions based on possibly aligned variable of a possibly aligned type,
calculate all alignments and sizes, check them against compiler if needed,
and check CPAchecker meets these numbers.

o-petrov committed:
Rewrite docstrings. Refactor string formatting for older python versions

o-petrov committed:
Fix docstrings in aligned_testing/graph.py

o-petrov committed:
Add arguments in test_aligned.py

Add arguments for graph options: loop depth, cycle depth, use plus for numbers,
use plus for pointers; check numbers or check pointers; don't run anything, just
generate .c files.

o-petrov committed:
Fix bug in test_aligned.py: get standard typeid for number type even if it was typedef'd

o-petrov committed:
test_aligned: implement prints, add option for alignments

1. Run program with prints and compare output with CPAchecker.
2. Make checking all alignments an option. By default check three alignments (default and two nearest).

o-petrov committed:
aligned_testing/graph.py: log edges (log when expressions are added)

o-petrov committed:
print graph stats: how many expressions there are in each node

o-petrov committed:
fix bug in aligned_testing/graph.py: mark plus as number or pointer arithmetics

o-petrov committed:
check void* with other pointers

o-petrov committed:
refactor test_aligned.py arguments

o-petrov committed:
add log option to test_aligned.py

o-petrov committed:
test_aligned: print model in .dot format

o-petrov committed:
Type options are not required for graph output

o-petrov committed:
Merge graph construction for numbers and pointers. Pointer to pointer is also possible.

o-petrov committed:
fix docstrings in graph.py

o-petrov committed:
Extract Graph from ExpressionGenerator. Allow multiple graphs. Construct graphs implicitly when needed

o-petrov committed:
make check_numbers() and check_pointers() one function

o-petrov committed:
add array support in graph construction

o-petrov committed:
fix align class for *...*v and small enhancements

o-petrov committed:
split generated text into a few files

o-petrov committed:
Declare more complicated variables besides numbers and pointers. Print graphs or nodes for types specified in this way instead of hardcoded types

o-petrov committed:
Add __int__ to integer literals to calculate size of arrays

o-petrov committed:
code style

o-petrov committed:
Print loops in .dot and node stats

o-petrov committed:
Add option to run checks on the model without running CPAchecker

o-petrov committed:
Fix adding loop expressions (no more dublicates)

o-petrov committed:
Add hash to expressions

o-petrov committed:
Remove init from CType.declare, as it is not used anyway

o-petrov committed:
Make typedefs properly

o-petrov committed:
Fix graphs printing for 0 cycle depth

o-petrov committed:
Drop loop depth for large batches of expressions

o-petrov committed:
Improve logging and output

Remove ending \n when logging declarations.
Use logger instead of logging.
Use nicks and alignment code to make typedefs names meaningful.
Remove redundant ';\n' after declarations in program texts.

o-petrov committed:
Move typenicks from machine and main script to CType

o-petrov committed:
Fix 2 bugs in generated programs

1. Print declarations of referenced types (pointed-to or elements of arrays).
2. Do not separate first two lines in its own file.

o-petrov committed:
Change asserts and prints in test programs

1. Add __str__ to C types.
2. Remove redundant asserts checking expression against title expression.
3. Print title expression and its type in asserts, expected values in prints.

o-petrov committed:
Replace HashMap with TreeMap. Add Alignment.comprareTo()

o-petrov committed:
Refactor MachineModel.getAlign()

Different cases require different default align calculations. All of these go after
there was no alignment found from attributes, so no need to pass default alignment
as parameter to getAlign()

o-petrov committed:
Add alignment for enums

Normally enum type can not have an alignment attribute. But variables of an enum type
still can have one, so elaborated type can have alignment. Also alignment can be forced
because of type canonization or such.

o-petrov committed:
Add alignment to composite types (not to members yet)

o-petrov committed:
Add alignment for members of not packed struct/union

o-petrov committed:
Store packed attribute in CComplexTypes

o-petrov committed:
If a type is of a packed member, flag it in Alignment

1. Store a boolean for 'packed struct/union member type alignment' in Alignment.
2. Update alignments of members when updating CCompositeType.
Also add Alignment.toString().

o-petrov committed:
Get packed attribute from AST and use it in size/alignment calculations

o-petrov committed:
Ignore tests with pointers with alignment attribute after asterisk

o-petrov committed:
Add alignment to bitfield type

o-petrov committed:
Ignore alignments for void types

o-petrov committed:
Add alignment to function type

o-petrov committed:
Remove default CType.getAlignment(), it is implemented in every subclass

o-petrov committed:
Fix aligned bit-fields offset etc.

Bit-field with aligned attribute is similar to 0-length bit-field and regular field.
Remove redundant sizeOfConsecutiveBitFields variable.
Use specified alignment if it is greater than default alignment of type of bit-field
(as for other fields of non-packed struct) when getting alignment of struct.

o-petrov committed:
Fix size of empty union

o-petrov committed:
Fix type of 'usual' bit-field

o-petrov committed:
Bit-fields are always consequent in packed structs

o-petrov committed:
Fix CBitfieldTest equals test: set distinct Alignment instances

o-petrov committed:
Add compatible integer type to enums

Do not default to signed int as compatible integer type for any enumerated type.

For each enumerator, its type is first of signed/unsigned int/long/long long
that can represent its value. (Types of enumerators may differ form each other
and from compatible type of the enum).

For enumerated type, its compatible integer type is first of signed/unsigned
[char/short]/int/long/long long that can represent values of all its enumerators.

Char and short can be a compatible type only for enum that has __packed__
attribute specified; enumerators of a packed enum are nonetheless at least int.

o-petrov committed:
replace assert with throw new AssertionError

o-petrov committed:
Check CSimpleType constructor usage

Add CONST_CHAR to CNumericTypes, use INT and CONST_CHAR instead of new instances
(they are immutable anyway), fix possible bug (copy alignment in handleModeAttribute()).

Druidos commented on src/org/sosy_lab/cpachecker/cfa/types/c/CEnumType.java:
```diff
@@ -63,6 +72,35 @@ public boolean isIncomplete() {
+   * @return integer type compatible with this enum
+   */
+  public CSimpleType getIntegerType() {
+    return integerType;
```
Require to add tests for enum with different internal types

o-petrov replied:
Added tests (and .xml for benchmark) for enums with one enumerator. Had to fix enumerator types too

Druidos commented on src/org/sosy_lab/cpachecker/cfa/ast/c/CIntegerLiteralExpression.java:
```diff
@@ -21,6 +21,7 @@ public class CIntegerLiteralExpression extends AIntegerLiteralExpression
   private static final long serialVersionUID = 7691279268370356228L;
   public static final CIntegerLiteralExpression ONE = createDummyLiteral(1L, CNumericTypes.INT);
   public static final CIntegerLiteralExpression ZERO = createDummyLiteral(0L, CNumericTypes.INT);
+  public static final CIntegerLiteralExpression THREE = createDummyLiteral(3L, CNumericTypes.INT);
```
Please move test value to corresponding CTypeToStringAlignmentTest

o-petrov replied:
Done

Druidos commented on src/org/sosy_lab/cpachecker/cfa/parser/eclipse/c/ASTConverter.java:
```diff
@@ -2579,47 +2810,117 @@ private CEnumType convert(IASTEnumerationSpecifier d) {
+      if (!enumerator.hasValue()) {
+        // some values might not have been simplified
+        // well, let it be int then
+        enumerator.setType(CNumericTypes.SIGNED_INT);
```
Please investigate whether this case is possible. Also please check if c.ASTConverter.convert(IASTEnumerationSpecifier.IASTEnumerator e, Long lastValue) correctly works especially with values close to ULLONG_MAX

o-petrov replied:
About ULLong values: It was not working correctly, but now I replaced long value field with BigInteger where needed.
I will check in which case enumerator does not have value later.

o-petrov replied:
About not simplified enumerator values: this case was possible; i hope i fixed it.

o-petrov committed:
Swap signed and unsigned integers in enum representations

GCC prefers unsigned types.

Druidos commented on src/org/sosy_lab/cpachecker/cfa/parser/llvm/CFABuilder.java:
```diff
@@ -1748,11 +1748,7 @@ private CExpression createGetElementPtrExp(final Value pItem, final Path pFileNa
-              true, false, CBasicType.CHAR, false, false, false, false, false, false, false);
-
-      CType stringType = new CPointerType(false, false, constCharType);
+      CType stringType = new CPointerType(false, false, CNumericTypes.CONST_CHAR);
```
Please check whether it is applicable for eclipse parser with pointers to strings

o-petrov replied:
What exactly? I've used an equal instance, and if it is changed somewhere, it is replaced with a new instance, so no other pointers are affected, as far as i understand it

o-petrov replied:
Replaced new CPointerType instance with a static one. (Still do not understand the question)

Druidos commented on src/org/sosy_lab/cpachecker/cfa/parser/llvm/LlvmTypeConverter.java:
```diff
@@ -284,13 +285,13 @@ private CType getSimplestCType(
         isComplex,
-        isImaginary,
-        isLongLong);
+        isImaginary, isLongLong);
```
Please fix style

o-petrov replied:
Done. Fixed all files

Druidos commented on src/org/sosy_lab/cpachecker/cfa/types/MachineModel.java:
```diff
@@ -1053,10 +1122,8 @@ public BigInteger calculateNecessaryBitfieldOffset(
   @Deprecated
   public BigInteger calculatePaddedBitsize(
       BigInteger pBitOffset,
-      BigInteger pSizeOfConsecutiveBitFields,
```
Please recheck whether this parameter is really unnecessary. Is it correctly process struct with several consecutive bitfields?

o-petrov replied:
As far as i've tested, it works fine. I had to add quite a bit of logic to the method, so i flatten it.

The reason for removing `sizeOfConsecutiveBitFields` is the following.
The variable stored consecutive bit-fields with padding on the edges, but all of this could be added to `bitOffset` in the first place. In a few methods `pSizeOfConsecutiveBitFields` is just added to the `bitOffset`, which hints the same.

o-petrov committed:
Add test programs for (not) packed enums with one enumerator

o-petrov committed:
Rewrite doc comments so they are true

o-petrov committed:
Use canonical types in enum compatible type candidate lists

o-petrov committed:
Two fixes getting underlying type of enum

o-petrov committed:
Store enumerator value as BigInteger, not long

C enumerators can be big values of unsigned long long type,
thus out of Java long bounds. Replace enumerator's long value field
with BigInteger one, fix getter method and its usage where possible.

o-petrov committed:
Fix types of enumerators

The type is not selected as smallest one from the list, it is
either int or the enum compatible type.

o-petrov committed:
Fix enumerator values not simplified

We need simplified enumerator values for correct enum compatible integer type,
even if 'cfa.simplifyConstExpressions' is disabled.

o-petrov committed:
Pad union size to its alignment

Fix two bugs: a) if all members have zero size, union also has zero size;
b) size of union has to be multiple of alignment, not 'not less than'.

o-petrov committed:
Fix alignment and offset of bit-fields

There are two use cases for getting alignment of a bit-field.
1. Computing alignment of containing struct;
2. Computing offsets and size of containing struct.

In first case, alignment attribute is applied as for a usual
field (can only increase alignment unless in packed struct).
Zero-length bit-filds are ignored while computing struct alignment.

In second case, alignment attribute is effective only if it
overaligns the type of the bit-field, or if the bit-field is not
zero-length and fits into alignment unit, or if the bit-field is
not zero-length and struct is packed.

o-petrov committed:
Move dummy constant 3 to the only test it is used in

o-petrov committed:
Fix style

o-petrov committed:
Fix style (multiline CSimpleType constructors)

Except for CNumericTypes file.

o-petrov committed:
Fix style (ASTConverter#handleAlignment comments)

o-petrov committed:
Fix pointer to const char public instances

Druidos commented on src/org/sosy_lab/cpachecker/cfa/parser/eclipse/c/ASTConverter.java:
```diff
@@ -2559,8 +2796,10 @@ private CCompositeType convert(IASTCompositeTypeSpecifier d) {
     List<CEnumerator> list = new ArrayList<>(d.getEnumerators().length);
-    Long lastValue = -1L; // initialize with -1, so the first one gets value 0
+    // initialize with -1, so the first one gets value 0
+    BigInteger lastValue = BigInteger.ONE.negate();
```
Better to use BitInteger.valueOf(-1)

o-petrov commented on src/org/sosy_lab/cpachecker/cfa/types/c/Alignment.java:
```diff
@@ -28,11 +30,20 @@ public class Alignment implements Serializable, Comparable<Alignment> {
+    varAligned = pVarAligned;
+    alignas = pAlignas;
+    insidePacked = pPacked;
   }
```
TODO call one constructor fom another

o-petrov committed:
Add test programs for aligned (bit-)members of structs and unions

o-petrov committed:
Two small fixes

o-petrov committed:
Rewrite and unify C types' to string methods where posssible

o-petrov committed:
Add warnings in ASTConverter#handleAlignment

including less aligned bit-fields, bit-fields with alignas, alignas reducing alignment

o-petrov committed:
Fix toString methods for pointer to struct in itself

o-petrov committed:
Fix C composite type to string: remove redundant comma

o-petrov committed:
Fix warning, expand comments

Warning is misleading when there is no alignment attribute.

o-petrov committed:
Fix recursion bug in #toASTString

If CPointerType references CCompositeType, always toASTString it
without any members (as elaborated type).
